zerver/lib/stream_subscription.py

```diff
@@ -313,3 +322,87 @@ def get_subscriptions_for_send_message(
         )
     )
     return query
+
+
+def update_user_streams_subscriber_count(
+    direction: Literal[1, -1], user_profile: UserProfile
+) -> None:
+    """
+    increment/decrement streams subscriber_count for a reactivated/deactivated user.
+
+    direction -> 1=increment, -1=decrement
+    """
+    get_user_subscribed_streams(user_profile).update(
+        subscriber_count=F("subscriber_count") + direction
+    )
+
+
+def update_streams_subscriber_count(
+    direction: Literal[1, -1],
+    streams: dict[int, set[int]],
+) -> None:
+    """increment/decrement streams subscriber_count for multiple users.
+
+    direction -> 1=increment, -1=decrement
+    """
+    if len(streams) == 0:
+        return
+
+    # list of tuples (stream_id, n_subscribers -- to be decremented/incremented)
+    # used as the columns of the temporary table delta_table.
+    stream_delta_values = [
+        (stream_id, len(subscribers) * direction) for stream_id, subscribers in streams.items()
+    ]
+
+    # The goal here is to update subscriber_count in a bulk efficient way,
+    # letting the database handle the deltas to avoid some race conditions.
+    # But unlike update_user_streams_subscriber_count which uses F() for a single delta value, we can't
+    # use F() to apply different deltas per row in a single update using ORM,
+    # so we use a raw SQL.
+    query = sql.SQL(
+        """UPDATE {stream_table}
+            SET subscriber_count = {stream_table}.subscriber_count + delta_table.delta
+            FROM (VALUES %s) AS delta_table(id, delta)
+            WHERE {stream_table}.id = delta_table.id;
+        """
+    ).format(stream_table=sql.Identifier(Stream._meta.db_table))
+
+    cursor = connection.cursor()
+    execute_values(cursor.cursor, query, stream_delta_values)
+
+
+@transaction.atomic(savepoint=False)
+def create_stream_subscription(
+    user_profile: UserProfile,
+    recipient: Recipient,
+    stream: Stream,
+    color: str = Subscription.DEFAULT_STREAM_COLOR,
+) -> None:
+    """
+    Creates a single stream Subscription object, incrementing stream.subscriber_count by 1
+    if user is active, in the same transaction.
+    """
+
+    # We only create a stream subscription in this function
+    assert recipient.type == Recipient.STREAM
+
+    Subscription.objects.create(
+        recipient=recipient,
+        user_profile=user_profile,
+        is_user_active=user_profile.is_active,
+        color=color,
+    )
+
+    if user_profile.is_active:  # nocoverage
+        Stream.objects.filter(id=stream.id).update(subscriber_count=F("subscriber_count") + 1)
+
+
+@transaction.atomic(savepoint=False)
+def bulk_create_stream_subscriptions(  # nocoverage
```

Comment:
This nocoverage also looks wrong.

Reply:
Yes

Reply:
Ok, I've add a test case for this, not directly calling the function, but instead ensuring it did it work.

But it's still missing from the test coverage report (i.e. false alarm) so we should keep the nocoverage.